internal/collector/processor/tail_sampling_processor.go

```diff
@@ -0,0 +1,429 @@
+			if decision, err := policy.Evaluator.OnDroppedSpans([]byte(traceID), trace); err != nil {
+				tsp.logger.Warn("OnDroppedSpans",
+					zap.String("policy", policy.Name),
+					zap.Int("decision", int(decision)),
+					zap.Error(err))
+			}
+		}
+	}
+}
+
+func prepareTraceBatch(spans []*tracepb.Span, singleTrace bool, batch *agenttracepb.ExportTraceServiceRequest) *agenttracepb.ExportTraceServiceRequest {
+	var traceBatch *agenttracepb.ExportTraceServiceRequest
+	if singleTrace {
+		// Special case no need to prepare a batch
+		traceBatch = batch
+	} else {
+		traceBatch = &agenttracepb.ExportTraceServiceRequest{
+			Node:     batch.Node,
+			Resource: batch.Resource,
+			Spans:    spans,
+		}
+	}
+	return traceBatch
+}
+
+// tTicker interface allows easier testing of ticker related functionality used by tailSamplingProcessor
+type tTicker interface {
+	// Start sets the frequency of the ticker and starts the perioc calls to OnTick.
+	Start(d time.Duration)
+	// OnTick is called when the ticker fires.
+	OnTick()
+	// Stops firing the ticker.
+	Stop()
+}
+
+type policyTicker struct {
+	ticker *time.Ticker
+	onTick func()
+}
+
+func (pt *policyTicker) Start(d time.Duration) {
+	pt.ticker = time.NewTicker(d)
+	go func() {
+		for range pt.ticker.C {
+			pt.OnTick()
+		}
+	}()
+}
+func (pt *policyTicker) OnTick() {
+	pt.onTick()
+}
+func (pt *policyTicker) Stop() {
+	pt.ticker.Stop()
+}
+
+var _ tTicker = (*policyTicker)(nil)
+
+// Variables related to metrics specific to queued processor.
+var (
+	tagPolicyKey, _ = tag.NewKey("policy")
+
+	statDecisionLatencyµs        = stats.Int64("sampling_decision_latency", "Latency (in microseconds) of a given sampling policy", "µs")
+	statOverallDecisionLatencyµs = stats.Int64("sampling_decision_timer_latency", "Latency (in microseconds) of each run of the sampling decision timer", "µs")
+
+	statTraceRemovalAgeSec           = stats.Int64("sampling_trace_removal_age", "Time (in seconds) from arrival of a new trace until its removal from memory", "s")
+	statLateSpanArrivalAfterDecision = stats.Int64("sampling_late_span_age", "Time (in seconds) from the sampling decision was taken and the arrival of a late span", "s")
+
+	statPolicyEvaluationErrorCount = stats.Int64("sampling_policy_evaluation_error", "Count of sampling policy evaluation errors", stats.UnitDimensionless)
+
+	statDroppedTooEarlyCount    = stats.Int64("sampling_trace_dropped_too_early", "Count of traces that needed to be dropped the configured wait time", stats.UnitDimensionless)
+	statNewTraceIDReceivedCount = stats.Int64("new_trace_id_received", "Counts the arrival of new traces", stats.UnitDimensionless)
+	statTracesOnMemoryGauge     = stats.Int64("sampling_traces_on_memory", "Tracks the number of traces current on memory", stats.UnitDimensionless)
+)
+
+// SamplingProcessorMetricViews return the metrics views according to given telemetry level.
+func SamplingProcessorMetricViews(level telemetry.Level) []*view.View {
```

Comment:
If we make a new package for sampling, can split these metrics into a `tail_sampling/metrics.go`

Reply:
done, looks better